src/core/comm/coll.cc

```diff
@@ -0,0 +1,389 @@
+  assert(MAX_NB_COMMS > 0);
+#ifdef LEGATE_USE_GASNET
+  int provided, init_flag = 0;
+  CHECK_MPI(MPI_Initialized(&init_flag));
+  if (!init_flag) {
+    CHECK_MPI(MPI_Init_thread(&argc, &argv, MPI_THREAD_MULTIPLE, &provided));
+  } else {
+    log_coll.print(
+      "Warning: MPI has been initialized by others, make sure MPI is initialized with "
+      "MPI_THREAD_MULTIPLE");
+  }
+  mpi_comms.resize(MAX_NB_COMMS, MPI_COMM_NULL);
+  for (int i = 0; i < MAX_NB_COMMS; i++) { CHECK_MPI(MPI_Comm_dup(MPI_COMM_WORLD, &mpi_comms[i])); }
+#else
+  thread_comms.resize(MAX_NB_COMMS);
+  for (int i = 0; i < MAX_NB_COMMS; i++) {
+    thread_comms[i].ready_flag = false;
+    thread_comms[i].buffers    = nullptr;
+    thread_comms[i].displs     = nullptr;
+  }
+#endif
+  coll_inited = true;
+  return CollSuccess;
+}
+
+int collFinalize(void)
+{
+  assert(coll_inited == true);
+  coll_inited = false;
+#ifdef LEGATE_USE_GASNET
+  for (int i = 0; i < MAX_NB_COMMS; i++) { CHECK_MPI(MPI_Comm_free(&mpi_comms[i])); }
+  mpi_comms.clear();
+  return MPI_Finalize();
+#else
+  for (int i = 0; i < MAX_NB_COMMS; i++) { assert(thread_comms[i].ready_flag == false); }
+  thread_comms.clear();
+  return CollSuccess;
+#endif
+}
+
+int collGetUniqueId(int* id)
+{
+  *id = current_unique_id;
+  current_unique_id++;
+  assert(current_unique_id <= MAX_NB_COMMS);
+  return CollSuccess;
+}
+
+#ifdef LEGATE_USE_GASNET
+MPI_Datatype dtypeToMPIDtype(CollDataType dtype)
+{
+  MPI_Datatype mpi_dtype = MPI_BYTE;
+  switch (dtype) {
+    case CollDataType::CollInt8: mpi_dtype = MPI_INT8_T; break;
+    case CollDataType::CollChar: mpi_dtype = MPI_CHAR; break;
+    case CollDataType::CollUint8: mpi_dtype = MPI_UINT8_T; break;
+    case CollDataType::CollInt: mpi_dtype = MPI_INT; break;
+    case CollDataType::CollUint32: mpi_dtype = MPI_UINT32_T; break;
+    case CollDataType::CollInt64: mpi_dtype = MPI_INT64_T; break;
+    case CollDataType::CollUint64: mpi_dtype = MPI_UINT64_T; break;
+    case CollDataType::CollFloat: mpi_dtype = MPI_FLOAT; break;
+    case CollDataType::CollDouble: mpi_dtype = MPI_DOUBLE; break;
+    default: log_coll.fatal("Unknown datatype"); assert(0);
+  }
+  return mpi_dtype;
+}
+
+static inline int match2ranks(int rank1, int rank2)
+{
+  // tag: seg idx + rank_idx + tag
+  // send_tag = sendto_global_rank * 10000 + global_rank (concat 2 ranks)
+  // which dst seg it sends to (in dst rank)
+  // recv_tag = global_rank * 10000 + recvfrom_global_rank (concat 2 ranks)
+  // idx of current seg we are receving (in src/my rank)
+  // example:
+  // 00 | 01 | 02 | 03
+  // 10 | 11 | 12 | 13
+  // 20 | 21 | 22 | 23
+  // 30 | 31 | 32 | 33
+  // 01's send_tag = 10, 10's recv_tag = 10, match
+  // 12's send_tag = 21, 21's recv_tag = 21, match
+
+  int tag;
+  constexpr int const max_ranks = 10000;
+  tag                           = rank1 * max_ranks + rank2;
+
+  // still under testing, will be used once if runs out of tags
```

Comment:
What testing are you still doing? If you're done with it, please remove these comments.

Reply:
Reply together with the previous 3 comments.
when max_ranks=1000, the max tag calculated by the equation is 1,000,100,000, while the max int32 is 2,147,483,647. If I set the max_ranks to 20000, the tag could be overflow. 
This is kind of like to have a hash function for pair<rank1, rank2>, I have tested several hash functions below, but not sure if they will have problems. So now I am just using the easiest one and if ranks > 10000, we can figure it out later.

Reply:
Do tags need to be unique for all pairs of logical ranks? I thought messages sent to rank k can use the same set of tags as those sent to rank k + 1, as their destination ranks would distinguish between those two sets of messages.

Reply:
Yes, because rank k and rank k+1 could be in the same node (MPI rank), actually what is why we need to use tag, while the original MPI implementation of collectives does not rely on tag. 
However, it is allowed sending to rank i and rank j to use the same pair as long as they are not in the same node, so I implemented a new way to compute the tag under the "still under testing"
`tag = rank1 % max_threads * max_ranks + rank2; `
I have also found 2 hash functions, and they are also under "still under testing", but they also have limitations.

Reply:
By "rank" I meant the MPI rank. The crux of the problem is that we want to distinguish between messages from MPI rank `i` on MPI rank `j`, where the messages are sent by different logical ranks on MPI rank `i`. What I was saying is that the tags we pass don't need to be globally unique but just need to be different for different threads on MPI rank `i`. Isn't this true? Or does the MPI require that the tags be globally unique for some reason I don't understand? Last time I checked with a toy example the former was the case. An MPI rank is already associated with every MPI send or receive request, so it feels redundant that the tag encodes it somehow.

Reply:
There's a possibility that the tags should be unique for all-to-all messages within the same MPI rank, but we can treat them specially.

Reply:
Let's go with the new tagging scheme. I agree that globally unique tags are easy to work with if you are still working on the communication routines, but I believe you already passed that point, so we better switch to a tagging scheme that would allow us to use practically infinitely many logical ranks.

Reply:
I have update the tagging scheme
https://github.com/eddy16112/legate.core/blob/coll/src/core/comm/coll.cc#L336